### ISSUE TITLE
Unclear "mismatched types" error

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-analyzer#18154]
As seen in this image, the inline error message is very unclear:  which type do I use, vs which type I **should be** using instead. It simply lists two types and their source files. It requires full dialog to be open to understand which is which.

![Image](https://github.com/user-attachments/assets/85ec2703-ae47-4bd0-af54-de6b04e44aee)

```rust
error[E0308]: mismatched types
   --> clippy_lints/src/functions/ref_option.rs:18:40
    |
18  |         if is_type_diagnostic_item(cx, *ty, sym::Option)
    |            -----------------------     ^^^ expected `rustc_middle::ty::Ty<'_>`, found `rustc_hir::Ty<'_>`
    |            |
    |            arguments to this function are incorrect
    |
    = note: `rustc_hir::Ty<'_>` and `rustc_middle::ty::Ty<'_>` have similar names, but are actually distinct types
note: `rustc_hir::Ty<'_>` is defined in crate `rustc_hir`
   --> /home/nyurik/.rustup/toolchains/nightly-2024-09-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_hir/src/hir.rs:2539:1
note: `rustc_middle::ty::Ty<'_>` is defined in crate `rustc_middle`
   --> /home/nyurik/.rustup/toolchains/nightly-2024-09-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_middle/src/ty/mod.rs:440:1
note: function defined here
   --> /home/nyurik/dev/rust/rust-clippy/clippy_utils/src/ty.rs:445:8
    |
445 | pub fn is_type_diagnostic_item(cx: &LateContext<'_>, ty: Ty<'_>, diag_item: Symbol) -> bool {
    |        ^^^^^^^^^^^^^^^^^^^^^^^
```

**rust-analyzer version**: `rust-analyzer version: 0.3.2112-standalone`

**rustc version**: `rustc 1.81.0 (eeb90cda1 2024-09-04)`

**editor or extension**: vscode, ext 0.3.2112

**relevant settings**: defaults

**repository link (if public, optional)**: (eg. [rust-analyzer](https://github.com/rust-lang/rust-analyzer))

**Comment by ChayimFriedman2** [comment]
The diagnostics get emitted by rustc in JSON, we have no control over them (although r-a native diagnostics are even worse currently).

**Comment by nyurik** [comment]
Makes sense, and rustc does make it clear because it uses comments on the code span itself - making it non-ambiguous. Does it make sense to show span info somewhere too in that error expansion?  Or would it require rustc's message to be altered to make it clearer which type is which?  Seems bad to loose span messages in r-a though...

**Comment by ChayimFriedman2** [comment]
We don't lose spans, we split the diagnostic so each part is shown on its span.

**Comment by nyurik** [comment]
I didn't see this message anywhere on the screen:

```
^^^ expected `rustc_middle::ty::Ty<'_>`, found `rustc_hir::Ty<'_>`
```

**Comment by ChayimFriedman2** [comment]
It should be somewhere there, but it could be a weak warning, rather invisible.

**Comment by Veykril** [comment]
VSCode doesn't do too well with "linked" diagnostics unfortunately, do note that there is a `Click for full compiler diagnostic` link in that hover message which will give you the full terminal rendered diagnostic in a new tab. So whenever you lack some information for a diagnostic in the editor view, clicking that will give you all the info you usually need.

**Comment by ChayimFriedman2** [comment]
As said we don't really have much choice here.
